runtime/io/io.cpp

```diff
@@ -0,0 +1,178 @@
+#include <gmp.h>
+#include <cstring>
+#include <stdexcept>
+
+#include "fcntl.h"
+#include "unistd.h"
+#include "runtime/alloc.h"
+#include "runtime/header.h"
+
+extern "C" {
+
+#define KCHAR char
+  
+  mpz_ptr move_int(mpz_t);
+
+  char * getTerminatedString(string * str) {
+    char * buf = (char *) malloc(sizeof(KCHAR) * (len(str) + 1));
+    strncpy(buf, str->data, len(str));
+    buf[len(str)] = '\0';
+    return buf;
+  }
+  
+  mpz_ptr hook_IO_open(string * filename, string * control) {
+    int flags = 0;
+    char * ctl = getTerminatedString(control);
```

Comment:
I don't think you need this call, as you aren't actually doing anything other than examining the characters of the string.